cpp/ql/test/experimental/query-tests/Security/CWE/CWE-125/semmle/tests/test.cpp

```diff
@@ -0,0 +1,194 @@
+typedef unsigned long size_t;
+#define MB_CUR_MAX 6
+#define MB_LEN_MAX 16
+int mbtowc(wchar_t *out, const char *in, size_t size);
+int wprintf (const wchar_t* format, ...);
+int strlen( const char * string );
+int checkErrors();
+
+void goodTest0()
+{
+  char * ptr = "123456789";
+  int ret;
+  int len;
+  len = 9;
+  for (wchar_t wc; (ret = mbtowc(&wc, ptr, len)) > 0; len-=ret) { // GOOD
+    wprintf(L"%lc", wc);
+  }
+}
+void goodTest1(const char* ptr)
+{
+  int ret;
+  int len;
+  len = strlen(ptr);
+  for (wchar_t wc; (ret = mbtowc(&wc, ptr, len)) > 0; len-=ret) { // GOOD
+    wprintf(L"%lc", wc);
+  }
+}
+void goodTest2(char* ptr)
+{
+  int ret;
+  ptr[10]=0;
+  int len = 9;
+  for (wchar_t wc; (ret = mbtowc(&wc, ptr, 16)) > 0; len-=ret) { // GOOD
```

Comment:
Should this be:
```suggestion
  int len = 10;
  for (wchar_t wc; (ret = mbtowc(&wc, ptr, len)) > 0; len-=ret) { // GOOD
```
(if not, what is the intent here?)

Reply:
in this and the next situation(goodTest2,goodTest3,goodTest4), there is an indication of the presence of zero at the end of the buffer, so the overflow will not occur. But if the read value is less than the maximum length of this character, then you can get unpredictable results, like badTest1,badTest2 here.
therefore it is necessary to specify some maximum character size value.

Reply:
I think I understand now, the rules / assumptions are:
 - if the input is not null terminated, we must specify the size of the input to prevent overflow.
 - if the input _is_ null terminated, overflow is not a concern because the `0` will prevent it.
   - but if the size parameter is less than `MB_LEN_MAX`, `mbtowc` could fail, so the size parameter should be >= `MB_LEN_MAX` (or `MB_CUR_MAX`).
 - the return value of `mbtowc` should only be used to update the input pointer, not the output pointer.

Reply:
thanks for the clear wording.
If I could write it right away, it would be easier to understand.
I'm sorry.